File: 0037-sudoku-solver/0037-sudoku-solver.cpp
class Solution {
private:
    bool row[9][9] = {false}, col[9][9] = {false}, box[9][9] = {false};
    vector<pair<int,int>> emptyCells;

    bool backtrack(vector<vector<char>>& board, int step) {
        if (step == emptyCells.size()) return true;

        auto [r,c] = emptyCells[step];
        int k = (r/3)*3+(c/3);

        for (int i=0; i<9; i++) {
            if (!row[r][i]&&!col[c][i]&&!box[k][i]) {
                board[r][c] = i+'1';
                row[r][i]=col[c][i]=box[k][i]=true;

                if (backtrack(board,step+1)) return true;

                board[r][c] = '.';
                row[r][i]=col[c][i]=box[k][i]=false;
            }
        }

        return false;
    }

public:
    void solveSudoku(vector<vector<char>>& board) {
        for (int i=0; i<9; i++) {
            for (int j=0; j<9; j++) {
                if (board[i][j] == '.') {
                    emptyCells.push_back({i,j});
                } else {
                    int ch = board[i][j] - '1';
                    int k = (i/3)*3+(j/3);

                    row[i][ch]=col[j][ch]=box[k][ch]=true;
                }
            }
        }

        backtrack(board,0);
    }
};

File: 0037-sudoku-solver/0037-sudoku-solver.py
class Solution(object):
    # TLE in Python
    # def isValid(self, board, ch, r, c):
    #     for i in range(len(board)):
    #         boxR = (r//3) * 3 + (i//3)
    #         boxC = (c//3) * 3 + (i%3)

    #         if (board[r][i]==ch or board[i][c]==ch or board[boxR][boxC]==ch):
    #             return False

    #     return True


    # def backtrack(self, board):
    #     for i in range(len(board)):
    #         for j in range(len(board[0])):
    #             if (board[i][j]=='.'):
    #                 for ch in '123456789':
    #                     if (self.isValid(board,ch,i,j)):
    #                         board[i][j]=ch
    #                         if (self.backtrack(board)):
    #                             return True
    #                         board[i][j]='.'
    #                 return False
    #     return True
    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: None Do not return anything, modify board in-place instead.
        """
        # self.backtrack(board)
        r = [set() for x in range(9)]
        c = [set() for x in range(9)]
        b = [set() for x in range(9)]
        empty = []

        #initialize sets + empties
        for i in range(9):
            for j in range(9):
                if board[i][j]!='.':
                    ch=board[i][j]
                    r[i].add(ch)
                    c[j].add(ch)
                    b[(i//3)*3+(j//3)].add(ch)
                else:
                    empty.append((i,j))

        def backtrack(idx=0):
            if idx == len(empty):
                return True
            
            i,j = empty[idx]
            k = (i//3) * 3 + (j//3)

            for ch in '123456789':
                if ch not in r[i] and ch not in c[j] and ch not in b[k]:
                    board[i][j]=ch
                    r[i].add(ch)
                    c[j].add(ch)
                    b[k].add(ch)

                    if (backtrack(idx+1)):
                        return True

                    board[i][j]='.'
                    r[i].remove(ch)
                    c[j].remove(ch)
                    b[k].remove(ch)
            return False

        backtrack()
